CocoonFs based persistence

nicstange opened the pull request:
This PR integrates CocoonFs for SVSM persistence.

I presented about CocoonFs in one of our devel meetings, see the [minutes from 2025-07-23](https://github.com/coconut-svsm/governance/blob/main/Meetings/devel-call-2025-07-23.md). To recap, it's been designed specifically with the persistence needs of TEEs in mind, the main features distinguishing it from existing solutions are
- Authentication of the FS state as a whole with a single, global Merkle tree -- other FSes implementing that are typically read-only.
- Fresh and random IVs at file granularity generated for every new write.
- Robustness against service disruptions by means of a journal.
For more details, see [the cocoonfs-cli README](https://github.com/coconut-svsm/cocoon-tpm/tree/main/cocoonfs-cli), which provides some further pointers to e.g. the spec.

One remark ahead: I believe it is well understood, but better to make it explicit anyway: as the secret key is obtained from the attestation subsystem, and that secret is not authenticated in any way yet, the overall scheme in its current state is prone to MITM attacks and **totally insecure**.

This PR implements the bare persistence functionality only, which culminates in providing `persistence_write_inode_sync()` and `persistence_read_inode_sync()` for use from the other SVSM code parts. I opted against wiring up the vtpm in this PR, because
- It's big enough by itself already.
- There are some security implications which need to be understood and documented, mainly due to the vtpm's identity not necessarily being unique anymore. We might end up implementing rollback protection at some point, but that would require talking to the outside and we don't have that yet. Until then, we might perhaps enforce `TPM_SU_CLEAR` for `TPM2_Shutdown()` for ensuring that the null hierarchy gets wiped at least.
- Lastly, @osteffenrh and @stefano-garzarella  have some patches for implementing vtpm persistence, c.f. [Oliver's kvm-forum-2025 branch](https://github.com/osteffenrh/coconut-svsm/commits/kvm-forum-2025/). It should be straight-forward to port these to the `persistence_{read,write}_inode_sync()` interface introduced here, and can help with that after coordinating with them.

As there are no persistence users yet, I added a small demo reading and incrementing a "boot counter" at startup. That can be found in the last commit, and is supposed to get reverted one way or the other at some point.

## Testing
- Setup attestation -- it's currently the only way to get a secret into the SVSM. Refer to `Documentation/docs/developer/ATTESTATION.md` in the sources.
- Enable the `cocoonfs` Cargo feature. This will pull in the `virtio-drivers` feature.
- Prepare a CocoonFs filesystem image using `cocoonfs-cli` (find it at the link provided above, or simply at crates.io):
  ```cocoonfs -i cocoonfs.img -f write-mkfs-info-header -H sha2 -C aes -t 128 -I 'ddeeff' -s 8M```
  Note that this merely writes a CocoonFs "filesystem creation info header", which doesn't require access to the key. The actual filesystem formatting will get done when the SVSM first attempts to open the filesystem.
- Attach the filesystem to the SVSM via the qemu command line (taken from `scripts/launch_guest.sh` in the sources):
  - append `,x-svsm-virtio-mmio=on` to the `-machine` spec and add
  - `-global virtio-mmio.force-legacy=false`
  - `-drive file=<path>/cocoonfs.img,format=raw,if=none,id=svsm_storage,cache=none`
  - `-device virtio-blk-device,drive=svsm_storage`

If everything goes well, you should see the following message at the first boot:
```
[SVSM] persistent CocoonFs storage opened successfully
[SVSM] persistence demo: no boot counter found yet
[SVSM] persistence demo: successfully wrote updated boot counter
```
and e.g.
```
[SVSM] persistent CocoonFs storage opened successfully
[SVSM] persistence demo: boot counter read back is 1
[SVSM] persistence demo: successfully wrote updated boot counter
```
etc. in subsequent ones.

nicstange committed:
crypto: manage crypto RNGs at a central place

Currently, the only user of a cryptographically secure RNG is
the attestation code. It uses a locally instantiated HashDrbg
implementation seeded from an X86RdSeed RNG, both provided by the
cocoon-tpm-crypto crate.

However, there are many more RNG users to come, and it is certainly
desirable to manage RNG instantiation at a central place:
- The initial RNG instantiation is not exactly a cheap operation,
  so it's worthwhile to maintain a pool of RNGs rather than to
  instantiate a new one for serving each and every request.
- The RNG instantiation procedure is is the only cocoon-tpm-crypto
  API primitive which is not agnostic of the configuration: the
  available RngCore implementations differ between the selectable
  crypto backends (pure Rust vs. boringssl). Right now the pure Rust
  one is used for the SVSM, but this will change.
- In the future we might want to stir the RNGs with some additional
  entropy obtained by some external means, entropy provided by an
  attestation server, from interrupt timings, whatnot.

Implement a minimal RNG pool for the SVSM. The new get_svsm_rng()
returns an exlusively owned SvsmRng instance, either from a (small)
pool if some is available, or a freshly instantiated one.
SvsmRng::drop() returns the instance back to the pool.

SvsmRng is an opaque type implementing the RngCore trait. Nothing else
must be assumed, other than it's smallish in size,
i.e. ~ sizeof(usize).

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
attest: use get_svsm_rng()

Make the attestation code to use the new get_svsm_rng() for obtaining
a RNG instance instead of open-coding the instantiation, so that it
will benefit from the SVSM's central RNG management infrastructure.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
kernel: make cocoon-tpm-crypto dep unconditional

Currently, the cocoon-tpm-crypto dependency is marked as optional
and enabled only by the "attest" Cargo feature.

However, more crypto users besides the attestation code will
undoubtedly get added to the SVSM over time, and gating the crypto
functionality by a disjunction of more and more Cargo features will
become impractical then.

It is my understanding that it's consensus among the SVSM development
community to use the cocoon-tpm-crypto crate for the crypto related
tasks, if possible, so just enable the dependency unconditionally.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
locking: implement get_mut() for RawSpinLock + RawRWLock

If an exclusive reference to the RawSpinLock or RawRWLock is available,
then access to the inner, guarded value may get granted without any
actual atomic locking operation.

Implement a get_mut() method for both, in line with
std::sync::Mutex::get_mut() and std::sync::RwLock::get_mut().

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
locking: implement From<T> for RawSpinLock<T> + RawRWLock<T>

From<T> for RawSpinLock<T> or RawRWLock<T> is trivial to implement and
provides a convenient way following standard Rust idioms for
constructing these. Implement From<T> for RawSpinLock<T> and
RawRWLock<T>.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
Add and integrate cocoon-tpm-utils-async dependency

The CocoonFs implementation builds heavily on the
cocoon-tpm-utils-async crate. Add it as a dependency.

The cocoon-tpm-utils-async crate defines the "SyncTypes" trait, which
enables users of synchronization primitives, such as the CocoonFs code,
to be completely generic over their respective implementation.
Introduce an implementation thereof suitable for the SVSM kernel
execution environment, "SvsmSyncTypes", plugging in the existing SVSM
SpinLock and RWLock implementations.

As some more sync/async infrastructure is to get added, do so in a new
"async" module.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
async: implement busy-polling Rust async task executor

Implement the rather trivial task_busypoll_to_completion() for
busy-polling Rust async tasks, typically represented by types
implementing the Future trait or similar, to completion.

For more flexibility, task_busypoll_to_completion() does not operate
on Rust Futures directly, but invokes a user provided closure that
takes only a Context. That closure might wrap an actual Future or,
alternatively, something else which is is only "Future-like".

Example:
  let result = task_busypoll_to_completion(
    |cx| { Future::poll(my_fut, cx) }
  );

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
Add cocoon-tpm-storage dep and wire it up to SVSM's BlockDriver

In working further towards supporting persistence built on CocoonFs,

- introduce a "persistence" Cargo feature for gating the functionality,
- add an optional dependency to the cocoon-tpm-storage crate,
  which provides the CocoonFs implementation,
- and wire it up to the SVSM's BlockDriver trait. More specifically,
  introduce a SvsmNvBlkDev struct wrapper around a generic D dereffing
  to BlockDriver, and implement the cocoon-tpm-storage's NvBlkDev trait
  for it.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
persistence: implement persistence_init() to open CocoonFs storage

CocoonFs is a special purpose filesystem designed for serving a
TEE's persistence needs: it brings strong security properties,
like e.g. global authentication with a Merkle tree and well randomized
encryptions, while also featuring a journal for robustness against
service disruptions. For more information, visit
https://github.com/coconut-svsm/cocoon-tpm/tree/main/cocoonfs-cli

Use CocoonFs for the SVSM's persistence.

In a first step, implement the initialization part. The initialization
is a split operation. In a first step the filesystem metadata is read
via the new persistence_discover(). That metadata may then serve as
input to the attestation procedure in order to obtain the key. Once
the key has been retrieved, it gets passed alongside the initially read
filesystem metadata to persistence_init() in order to conclude the
initialization. Note that in case a special CocoonFs "filesystem
creation info header"  was found by the prior persistence_discover(),
then the filesystem will get transparently formatted for use with the
supplied key.

Currently, the SVSM supports only a single block device, the
BLOCK_DEVICE, and persistence_discover() examines only that
accordingly. Once the CocoonFs filesystem got opened successfully by
persistence_init(), it will assign a corresponding CocoonFs instance to
the new global SVSM_COCOONFS_INSTANCE for subsequent use.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
persistence: implement inode read + write primitives

Implement persistence_write_inode_sync() and
persistence_read_inode_sync() for general use from other SVSM code.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange committed:
persistence: implement a demo

Implement a small demo for the persistence functionality.

At every SVSM startup, a "boot counter" is read from inode number 16,
incremented by one and written back.

Expected output at first boot:
  [SVSM] persistent CocoonFs storage opened successfully
  [SVSM] persistence demo: no boot counter found yet
  [SVSM] persistence demo: successfully wrote updated boot counter

At second boot:
  [SVSM] persistent CocoonFs storage opened successfully
  [SVSM] persistence demo: boot counter read back is 1
  [SVSM] persistence demo: successfully wrote updated boot counter

At third boot:
  [SVSM] persistent CocoonFs storage opened successfully
  [SVSM] persistence demo: boot counter read back is 2
  [SVSM] persistence demo: successfully wrote updated boot counter

And so on.

The added code is only intended for demonstration purposes,
and expected to get removed again once there are real users
of the persistence functionality.

Signed-off-by: Nicolai Stange <nstange@suse.de>

osteffenrh reviewed (changes requested):
Here are some initial comments. I'll do some more work later.

tylerfanelli commented:
> > > > A higher level comment: Maybe we can agree on how to encode secrets and binary data across the project. Here hex is used, while the kbs uses base64 (also @tylerfanelli)
> > > 
> > > 
> > > The attestation subsystem uses base64 because that's what is returned by a standard KBS. I could modify the KBS `aproxy` module to convert everything to hex, however. Whichever is preferred.
> > 
> > 
> > I believe @osteffenrh's comment was exclusively about CLIs and user experience. That is, `kbs-test`'s `--secret`, which expects base64 encoding vs. `cocoonfs-cli`'s `-K` which takes the very same value in hex. For anything where humans are not involved, base64 is clearly more compact, so I would personally stick with that. For CLIs OTOH, my gut feeling is that specifying binary data as hex is more common in general, but I might be wrong about that. If nothing else, hex allows for figuring the key length more or less right from eyeing at the string. So I'm leaning towards hex, but would absolutely be open for using base64 in case that's more established in the CoCo ecosystem.
> 
> Yes, I meant the human-facing representations.

Oh, I see. Yes, I agree that hex encoding would be preferred.

nicstange commented on kernel/src/attest.rs:
```diff
@@ -243,27 +239,8 @@ impl From<AttestationError> for SvsmError {
-
+    let mut rng = get_svsm_rng()?;
     let curve_ops = curve.curve_ops()?;
-
```
Not really IMO: before this change the rng instantiation code in `sc_key_generate()` had been non-trivial and it made sense to separate it visually from the rest. Now both the `rng` + `curve_ops` input parameter instantiations for the subsequent `EccKey::generate()` (missing from the diff context unfortunately) are mere one-liners and it would IMO look a bit odd to separate them by a blank line from `EccKey::generate()`. Do you agree or shall I still restore the blank line?

nicstange replied:
Will do, thanks!

osteffenrh commented on kernel/src/persistence.rs:
```diff
@@ -694,3 +697,216 @@ pub fn persistence_init(mut key: Zeroizing<Vec<u8>>) -> Result<(), SvsmError> {
+#[allow(unused)]
+pub fn persistence_write_inode_sync(
+    inode: u32,
+    data: Zeroizing<Vec<u8>>,
```
Why do we need the data to be written to be in a Zeroizing buffer?

stefano-garzarella commented on kernel/src/persistence.rs:
```diff
@@ -0,0 +1,564 @@
+    }
+}
+
+impl NvChip for SvsmBlockDriverNvChip {
```
OKay, the name is a bit confusing IMO, it seems related to TPM's NvData.

nicstange committed:
Documentation: add developer doc for persistence functionality

Add developer documentation for the persistence subsystem.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange commented:
> For moving the convenience helpers into the cocoon-tpm-storage crate, I would not be against doing that in some generalized form (I'd like to keep the async executor out of that crate, so it would come in `Future` form).

When managing the UEFI var dir in a dedicated file, one would probably want to combine updates to that and to the some UEFI var files in a single CocoonFs transaction though, for atomicity.

kraxel commented:
> When managing the UEFI var dir in a dedicated file, one would probably want to combine updates to that and to the some UEFI var files in a single CocoonFs transaction though, for atomicity.

Goes into the 'nice to have' bucket, but it is not a hard requirement.

* When adding a variable I can write the variable file first and update the index second.
* When removing a variable I can update the index first and remove the variable file second.
* When updating a variable I do not need to touch the index in the first place.

That should make sure things are consistent for the variable store, even though there is the risk that it leaves variable files in the filesystem which are not referenced in the index in case only one of the two updates is successfully flushed to the filesystem for add/remove ops.

nicstange committed:
Upgrade cocoon-tpm-utils-common, -tpm2-interface + -crypto deps

Increment the minimum versions in the Cargo.toml and run
  $ cargo update --recursive cocoon-tpm-utils-common \
        cocoon-tpm-tpm2-interface                    \
        cocoon-tpm-crypto
to take care of Cargo.lock.

Signed-off-by: Nicolai Stange <nstange@suse.de>

nicstange commented on Documentation/docs/developer/PERSISTENCE.md:
```diff
@@ -0,0 +1,100 @@
+  skipped.
+* If no working CocoonFs image had been found yet, and the next block device has
+  a "CocoonFs creation header" on it, the filesystem would get formatted with
+  whatever key had been obtained from attestion and henceforth be used for the
```
Applied, thanks!

luigix25 commented on kernel/src/persistence.rs:
```diff
@@ -0,0 +1,613 @@
+}
+
+impl<D: ops::Deref<Target: BlockDriver> + Send + Sync + Unpin> SvsmNvBlkDev<D> {
+    fn new(driver: D) -> Self {
```
i ran `make clippy` on every single commit and it fails in "Add cocoon-tpm-storage dep and wire it up to SVSM's BlockDriver":

```
error: associated function `new` is never used
  --> kernel/src/persistence.rs:30:8
   |
29 | impl<D: ops::Deref<Target: BlockDriver> + Send + Sync + Unpin> SvsmNvBlkDev<D> {
   | ------------------------------------------------------------------------------ associated function in this implementation
30 |     fn new(driver: D) -> Self {
   |        ^^^
   |
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`
```

nicstange replied:
Annotated `fn new()` with a `allow(dead_code)`  and removed that again in the subsequent commit introducing an actual usage.

luigix25 commented on kernel/src/persistence.rs:
```diff
@@ -0,0 +1,1123 @@
+        }
+
+        for i in 0..this.io_blocks_count {
+            let dev_block_id = (this.io_block_index + i) >> dev_io_block_size_128b_log2;
```
claude is reporting that this shift is unnecessary:

```
In SvsmNvBlkDevTrimFuture::poll (line 608), io_block_index is the NvBlkDev io-block index — confirmed by the bounds check on line 590–591, which compares it directly against dev_io_blocks (the total io-block
  count from io_blocks()):

  } else if self.io_block_index > dev_io_blocks          // io_block_index in io-block units
      || dev_io_blocks - self.io_block_index < self.io_blocks_count

  But then line 608 applies the same shift that was correct for 128-byte-unit values:

  let dev_block_id = (this.io_block_index + i) >> dev_io_block_size_128b_log2;
  //                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^  already in io-block == dev-block units
  //                                               the shift makes no sense here

  Since one io-block already equals one dev-block, the correct formula is just:

  let dev_block_id = this.io_block_index + i
```

On a quick look it kind of makes sense, but I'm not 100% sure

nicstange replied:
Claude was right, fixed. Thanks a lot for running it!

luigix25 commented on kernel/src/persistence.rs:
```diff
@@ -0,0 +1,1123 @@
+    let data = Zeroizing::new(boot_counter.to_le_bytes().to_vec());
+    match persistence_write_inode_sync(SvsmPersistenceStaticInode::Demo as u64, data, true) {
+        Ok(_) => log::info!("persistence demo: successfully wrote updated boot counter"),
+        Err(e) => log::error!("persistence demo: boot counter updating write failed: {e:?})"),
```
nit: there is an extra `)`

nicstange replied:
Fixed.

luigix25 commented on Documentation/docs/developer/PERSISTENCE.md:
```diff
@@ -0,0 +1,100 @@
+with the key obtained from attestation procedure. If created with a key, the
+filesystem will get formatted right away at `cocoonfs-cli` invocation time.
+
+At startup, the SVSM would iterate over all block devices attach to it and use
```
/s/attach/attached?

nicstange replied:
Applied.

luigix25 commented on kernel/src/crypto/rng.rs:
```diff
@@ -0,0 +1,147 @@
+/// Instances of `SvsmRng` are to be obtained through [`get_svsm_rng()`].
+///
+/// No assumptions must be made about `SvsmRng`, other than instances thereof are small, typically
+/// of pointer size), and that it implements [`RngCore`].
```
nit: extra ), or missing (

nicstange replied:
Fixed.

luigix25 commented on kernel/src/persistence.rs:
```diff
@@ -0,0 +1,1123 @@
+    }
+}
+
+/// Convert a [`SvsmError` to a [`NvBlkDevIoError`].
```
missing ]

nicstange replied:
Added.

luigix25 commented on kernel/src/persistence.rs:
```diff
@@ -0,0 +1,1123 @@
+/// * [`persistence_init()`]
+pub fn persistence_discover() -> Result<Option<PersistenceBootstrapInfo>, SvsmError> {
+    // Inspect the global BLOCK_DEVICE and see if there's a CocoonFs on it, either already formatted
+    // or one with with a mkfsinfo header, which will get formatted transparently at first
```
extra with

nicstange replied:
Fixed.

luigix25 commented on kernel/src/persistence.rs:
```diff
@@ -0,0 +1,1123 @@
+}
+
+impl PersistenceBootstrapInfo {
+    /// Access the matadata.
```
/s/matadata/metadata?

nicstange replied:
Applied.

luigix25 commented on kernel/src/async.rs:
```diff
@@ -0,0 +1,133 @@
+// SAFETY: The RawWakerVTable signature requires the function to be marked unsafe, there's nothing
+// actually unsafe about it -- data is always a null ptr and not accessed.
+unsafe fn nop_raw_waker_wake_by_ref(_data: *const ()) {
+    // Do nothing -- it's a nop waker.
```
why not adding an assert like in the other cases?

nicstange replied:
For context, the assert in the other cases is `debug_assert!(data.is_null())`.

For the answer: the other cases have that, because the correctness of their implementations (to do nothing) depends on the `data` argument being a null pointer -- they are required to manage the storage backing `data` one way or the other (clone, drop, consume). C.f. https://doc.rust-lang.org/std/task/struct.RawWakerVTable.html for reference. While it is true that `data` is null in the `wake_by_ref()` case also, nothing relies on it, and that's why I omitted the `assert` from there.

luigix25 replied:
ok thanks!

luigix25 commented on kernel/src/locking/spinlock.rs:
```diff
@@ -7,6 +7,7 @@
 use super::common::*;
 use crate::types::TPR_LOCK;
 use core::cell::UnsafeCell;
+use core::convert::From;
```
i think this can be dropped

nicstange replied:
Indeed, dropped it.

luigix25 commented on kernel/src/locking/rwlock.rs:
```diff
@@ -7,6 +7,7 @@
 use super::common::*;
 use crate::types::TPR_LOCK;
 use core::cell::UnsafeCell;
+use core::convert::From;
```
ditto

nicstange replied:
Dropped as well.

luigix25 commented on kernel/src/crypto/rng.rs:
```diff
@@ -0,0 +1,147 @@
+impl SvsmRng {
+    fn instantiate() -> Result<Self, CryptoError> {
+        let mut hash_drbg_entropy = zeroize::Zeroizing::new(
+            FixedVec::<u8, 5>::new_with_default(HashDrbg::min_seed_entropy_len(
```
i think 5 is log2(32), I wonder if we should use a const

nicstange replied:
Introduced local `const DRBG_HASH_ALG_DIGEST_LEN_LOG2`, with a a value derived from the `const SVSM_RNG_DRBG_HASH_ALG`.

luigix25 commented on kernel/src/crypto/rng.rs:
```diff
@@ -0,0 +1,147 @@
+/// to take a currently unused instance from the pool before attempting to instantiate one.
+#[allow(clippy::type_complexity)]
+static SVSM_RNG_POOL: SpinLock<
+    [Option<Box<rng::ChainedRng<rng::X86RdSeedRng, rng::HashDrbg>>>; 4],
```
adding const like `SVSM_RNG_POOL_SIZE` here?

nicstange replied:
Done.

luigix25 commented on kernel/src/svsm.rs:
```diff
@@ -548,11 +550,22 @@ fn svsm_init(launch_info: &KernelLaunchInfo) {
 
-        // Nothing to do with data at the moment, simply print a success message.
+        let mut proxy = AttestationDriver::try_from(Tee::Snp).unwrap();
+        let secret = proxy.attest().unwrap();
```
what about using expect with an error message?

nicstange replied:
Note that it's been like that before this PR and the attestation subsystem would have emitted some warnings on what went wrong in case the `unwrap()`  here panicked. This code here is tentative and will get reworked anyway, so I'd simply leave it as-is for now.

luigix25 replied:
sure, fine for me!

luigix25 reviewed (commented):
Before merging I suspect the demo part should be dropped.

*Huge* thanks for this work!

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -551,11 +553,21 @@ fn svsm_init(launch_info: &KernelLaunchInfo) {
 
-        // Nothing to do with data at the moment, simply print a success message.
+        let mut proxy = AttestationDriver::try_from(Tee::Snp).unwrap();
+        let secret = proxy.attest().unwrap();
```
IIUC as next step (I guess with @armenon-rh helps since one of his PRs) the `persistence_bootstrap_info` should be passed to the `proxy.attest()`, right?

nicstange replied:
Correct. It provides access to the auxiliary fs metadata, which, for our purposes, is supposed to contain a wrapped key.

stefano-garzarella replied:
@nicstange great, thanks for confirming!

stefano-garzarella commented on kernel/Cargo.toml:
```diff
@@ -95,7 +95,7 @@ attest = [
 
 default = []
-cocoonfs = ["dep:cocoon-tpm-storage", "block", "virtio-drivers"]
+cocoonfs = ["dep:cocoon-tpm-storage", "block", "virtio-drivers", "attest"]
```
I'm confused here, if `cocoonfs` depends on `attest`, why in `kernel/src/svsm.rs` we have `#[cfg(not(feature = "cocoonfs"))]` inside the attestation block? (I mean here: https://github.com/coconut-svsm/svsm/pull/806/changes/c9c5317690a98c5225cc918b163c315e354e8ec4#diff-b3e68fb174065a98fa188eecccb858b06bb0a61f3fa8dc20a5194b7e6a4f9666R565)

nicstange replied:
The `#[cfg(not(feature = "cocoonfs"))]` branch is basically the original code retained, for the likely case someone would like to work on attestation without having the whole persistence thing setup/enabled.

stefano-garzarella replied:
oh, I see now, my mind was thinking the opposite...

stefano-garzarella commented on kernel/src/svsm.rs:
```diff
@@ -567,6 +567,7 @@ fn svsm_init(launch_info: &KernelLaunchInfo) {
         #[cfg(feature = "cocoonfs")]
         if let Some(persistence_bootstrap_info) = persistence_bootstrap_info {
             persistence_init(persistence_bootstrap_info, &secret).unwrap();
+            persistence_demo();
```
I'm fine to keep the demo here, maybe just add a TODO comment to remove it, or a special tmp feature to enable/disable it (not strong opinion on that, I'm fine to keep it till we don't use it for attestation)

stefano-garzarella replied:
of course, feel free to ignore this if you would like to remove the demo commit from this PR.

nicstange replied:
I added a TODO comment right before the `persistent_demo()` invocation.

stefano-garzarella commented on kernel/src/async.rs:
```diff
@@ -60,3 +61,73 @@ impl sync_types::SyncTypes for SvsmSyncTypes {
+// SAFETY: The RawWakerVTable signature requires the function to be marked unsafe, there's nothing
+// actually unsafe about it -- data is always a null ptr and not accessed.
+unsafe fn nop_raw_waker_clone(data: *const ()) -> task::RawWaker {
+    // No ressources are associated with the nop waker.
```
nit:
```suggestion
    // No resources are associated with the nop waker.
```

stefano-garzarella replied:
Copied&Pasted in the next 2 functions as well

nicstange replied:
The code got removed now (in favor of `task::Waker::noop()`).

stefano-garzarella commented on kernel/src/async.rs:
```diff
@@ -60,3 +61,73 @@ impl sync_types::SyncTypes for SvsmSyncTypes {
+pub fn task_busypoll_to_completion<O, T: FnMut(&mut task::Context<'_>) -> task::Poll<O>>(
+    mut task: T,
+) -> O {
+    let raw_waker = nop_raw_waker_new();
```
Can we use https://doc.rust-lang.org/core/task/struct.Waker.html#method.noop ?

IIUC it's stable since 1.85 and we are at 1.86, so should be fine and allow us to remove some code here. WDYT?

nicstange replied:
Hah! Nice, thanks for the pointer, will look into it.

nicstange replied:
And it's working great!

stefano-garzarella commented on kernel/src/persistence.rs:
```diff
@@ -817,3 +820,257 @@ pub fn persistence_init(
+/// * `data` - The data to write to `inode`.
+/// * `issue_sync` - Whether or not to issue a sync request to the underlying storage after the
+///   write has completed. That's best effort though and relies on the host to behave well.
+#[allow(unused)]
```
Could this be removed after next commit that adds the demo?

nicstange replied:
Let's keep it until it's been decided whether the demo should get included or not.

stefano-garzarella replied:
yeah, I saw that comment later. Makes sense to me!

stefano-garzarella commented on kernel/src/persistence.rs:
```diff
@@ -817,3 +820,257 @@ pub fn persistence_init(
+/// # Arguments:
+///
+/// * `inode` - Number of the inode whose data to read.
+#[allow(unused)]
```
Ditto.

stefano-garzarella commented on kernel/src/persistence.rs:
```diff
@@ -1074,3 +1074,49 @@ pub fn persistence_read_inode_sync(inode: u64) -> Result<Option<Zeroizing<Vec<u8
+        }
+    };
+
+    boot_counter += 1;
```
nit (feel free to ignore, it's a demo xD):
```suggestion
    boot_counter = boot_counter.wrapping_add(1);
```

nicstange replied:
Applied :)

stefano-garzarella commented on kernel/Cargo.toml:
```diff
@@ -94,6 +95,7 @@ attest = [
 attest-serial = ["attest"]
 
 default = []
+cocoonfs = ["dep:cocoon-tpm-storage", "block", "virtio-drivers"]
```
Not a strong opinion, but would it better to change the name of the feature, I mean something like `persistence` like the module we are adding. Anyway, `cocoonfs` is also fine, just something that came to mind while reviewing.

nicstange replied:
Way back we discussed the possibility to support other filesystems in the future and I didn't want to reserve the generic "persistence" now. But either way is fine for me, awaiting instructions ;)

stefano-garzarella replied:
IMHO we can have "persistence" that for now uses cocoonfs since it's the only one we support. When we will add a new one, we can maybe add more features to enable/disable each one.

nicstange replied:
Renamed the "cocoonfs" Cargo feature to "persistence".

stefano-garzarella reviewed (commented):
@nicstange thanks for the huge work! Overall LGTM. I left some minor comments (mostly for my understanding) and some nits (feel free to ignore them and close those comments).

I think we are near to merge this :-)

osteffenrh commented on kernel/src/persistence.rs:
```diff
@@ -694,3 +697,216 @@ pub fn persistence_init(mut key: Zeroizing<Vec<u8>>) -> Result<(), SvsmError> {
+///
+/// * `inode` - Number of the inode whose data to read.
+#[allow(unused)]
+pub fn persistence_read_inode_sync(inode: u32) -> Result<Option<Zeroizing<Vec<u8>>>, SvsmError> {
```
Let's leave it as is. We can still improve the zeroing later, if needed. This is not a big problem.

Thanks for your work!
